internal/hcsoci/resources.go

```diff
@@ -34,151 +32,82 @@ type Resources struct {
 
-type scsiMount struct {
-	// path is the host path to the vhd that is mounted.
-	path string
-	// autoManage if `true` means that on cleanup, the runtime should
-	// automatically delete this vhd.
-	autoManage bool
+// ResourceCloser is a generic interface for the releasing of a resource. If a resource implements
+// this interface(which they all should), freeing of that resource should entail one call to
+// <resourceName>.Release(ctx, all)
+type ResourceCloser interface {
+	Release(context.Context) error
 }
 
-// TODO: Method on the resources?
-func ReleaseResources(ctx context.Context, r *Resources, vm *uvm.UtilityVM, all bool) error {
-	if vm != nil && r.addedNetNSToVM {
-		if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
-			log.G(ctx).Warn(err)
-		}
-		r.addedNetNSToVM = false
-	}
+// AutoManagedVHD struct representing a VHD that will be cleaned up automatically.
+type AutoManagedVHD struct {
+	hostPath string
+}
 
-	if r.createdNetNS {
-		for len(r.networkEndpoints) != 0 {
-			endpoint := r.networkEndpoints[len(r.networkEndpoints)-1]
-			err := hns.RemoveNamespaceEndpoint(r.netNS, endpoint)
-			if err != nil {
-				if !os.IsNotExist(err) {
-					return err
-				}
-				log.G(ctx).WithFields(logrus.Fields{
-					"endpointID": endpoint,
-					"netID":      r.NetNS(),
-				}).Warn("removing endpoint from namespace: does not exist")
-			}
-			r.networkEndpoints = r.networkEndpoints[:len(r.networkEndpoints)-1]
-		}
-		r.networkEndpoints = nil
-		err := hns.RemoveNamespace(r.netNS)
-		if err != nil && !os.IsNotExist(err) {
-			return err
-		}
-		r.createdNetNS = false
+// Release removes the vhd.
+func (vhd *AutoManagedVHD) Release(ctx context.Context) error {
+	if err := os.Remove(vhd.hostPath); err != nil {
+		log.G(ctx).WithError(err).Warnf("failed to remove automanage-virtual-disk at: %q", vhd.hostPath)
 	}
+	return nil
+}
 
-	if vm != nil && all {
-		for len(r.vsmbMounts) != 0 {
-			mount := r.vsmbMounts[len(r.vsmbMounts)-1]
-			if err := vm.RemoveVSMB(ctx, mount); err != nil {
-				return err
-			}
-			r.vsmbMounts = r.vsmbMounts[:len(r.vsmbMounts)-1]
-		}
-
-		for len(r.pipeMounts) != 0 {
-			mount := r.pipeMounts[len(r.pipeMounts)-1]
-			if err := vm.RemovePipe(ctx, mount); err != nil {
-				return err
+// ReleaseResources releases/frees all of the resources associated with a container. This includes
+// Plan9 shares, vsmb mounts, pipe mounts, network endpoints, scsi mounts, vpci devices and layers.
+func ReleaseResources(ctx context.Context, r *Resources, vm *uvm.UtilityVM, all bool) error {
+	if vm != nil {
+		if r.addedNetNSToVM {
+			if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
+				log.G(ctx).Warn(err)
 			}
-			r.pipeMounts = r.pipeMounts[:len(r.pipeMounts)-1]
+			r.addedNetNSToVM = false
 		}
 
-		for len(r.plan9Mounts) != 0 {
-			mount := r.plan9Mounts[len(r.plan9Mounts)-1]
-			if err := vm.RemovePlan9(ctx, mount); err != nil {
-				return err
+		// cleanup container state
+		if vm.DeleteContainerStateSupported() {
```

Comment:
We should be careful to not reorder the cleanup operations as part of the refactoring. It looks like this was done later previously.

Reply:
Mostly done to avoid two different if vm != nil blocks and no behavior changed while testing but I agree. I'll move it back to how it was originally structured